guide/kong_integration.md

```diff
@@ -0,0 +1,196 @@
+# Using Kong as an API gateway for OpenFaaS
+
+[Kong](https://getkong.org) is an API gateway that provides features such as security, logging, and rate limiting. By putting this in front of OpenFaaS you can quickly get access to these things and a lot more via [the many other plugins written](https://getkong.org/plugins/) for it.
+
+Below is a demo of how you could use Kong as an authentication layer for OpenFaaS.
+
+## Setup OpenFaaS
+
+```
+$ docker swarm init --advertise-addr $(hostname -i)
+
+$ git clone https://github.com/alexellis/faas && \
+  cd faas && \
+  ./deploy_stack.sh
+```
+
+Check that one of the sample functions works
+
+```
+$ curl localhost:8080/function/func_echoit -d 'hello world'
+hello world
+```
+
+## Setup Kong
+```
+$ docker service create --network func_functions --detach=false \
+      --name kong-database \
+      -p 5432:5432 \
+      -e "POSTGRES_USER=kong" \
+      -e "POSTGRES_DB=kong" \
+      postgres:9.4
+
+$ docker service create --network func_functions --detach=false \
+      --restart-condition=none --name=kong-migrations \
+      -e "KONG_DATABASE=postgres" \
+      -e "KONG_PG_HOST=kong-database" \
+      kong:latest kong migrations up
+
+$ docker service create --network func_functions --name kong \
+      -e "KONG_DATABASE=postgres" \
+      -e "KONG_PG_HOST=kong-database" \
+      -e "KONG_PROXY_ACCESS_LOG=/dev/stdout" \
+      -e "KONG_ADMIN_ACCESS_LOG=/dev/stdout" \
+      -e "KONG_PROXY_ERROR_LOG=/dev/stderr" \
+      -e "KONG_ADMIN_ERROR_LOG=/dev/stderr" \
+      -p 8000:8000 \
+      -p 8443:8443 \
+      -p 8001:8001 \
```

Comment:
Worth adding a short firewall section to highlight all the admin/api ports that should be restricted in one place perhaps?

Reply:
Is it possible to run with fewer ports exposed?

Reply:
It is possible. Of the bat I can think of 2 ways of doing this.
Lets suppose we only want to expose 8443 and not 8000 (kong http user port), or 80001 (kong admin port)
1. on the host machine, get the IP of kong and use that to make all the admin requests to kong.
2. instead of making the admin requests from the host, we can docker exec into the kong machine and do all the updates from there.

However, I do think this makes the readme more confusing and seems to diverge a bit from how kong is documented. If the user is really going to put this in prod, they should have some sort of firewall regardless, no? You dont want to expose your entire machine to the network.

Also, if you are concerned with exposing ports, we should also hide the OpenFaaS ports as well since only Kong really needs to know them.